payments/models.py

```diff
@@ -154,3 +169,309 @@ def __str__(self):
+    status = EnumField(OrderStatus, default=OrderStatus.WAITING)
+    comment = models.TextField(blank=True, null=True)
+    base_price = models.DecimalField(
+        verbose_name=_("price"),
+        max_digits=10,
+        decimal_places=2,
+        validators=[MinValueValidator(Decimal("0.01"))],
+        blank=True,
+        null=True,
+    )
+    tax_percentage = models.DecimalField(
+        verbose_name=_("tax percentage"),
+        max_digits=5,
+        decimal_places=2,
+        blank=True,
+        null=True,
+    )
+    due_date = models.DateField(
+        verbose_name=_("due date"), default=calculate_order_due_date,
+    )
+
+    _product_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="product",
+    )
+    _product_object_id = models.UUIDField(
+        null=True, blank=True, verbose_name=_("product")
+    )
+
+    _lease_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="lease",
+    )
+    _lease_object_id = models.UUIDField(null=True, blank=True, verbose_name=_("lease"))
+
+    objects = OrderManager()
+
+    def __str__(self):
+        return f"{self.product} [{self.status}]"
+
+    @property
+    @rounded
+    def full_price(self):
+        return convert_pretax_to_aftertax(self.base_price, self.tax_percentage)
+
+    def clean(self):
+        if self.product and not isinstance(
+            self.product, (BerthProduct, WinterStorageProduct)
+        ):
+            raise ValidationError(_("You cannot assign other types of products"))
+
+        if self.lease and not isinstance(self.lease, (BerthLease, WinterStorageLease)):
+            raise ValidationError(_("You cannot assign other types of leases"))
+
+        if self.product and self.lease:
+            if isinstance(self.product, BerthProduct) and not isinstance(
+                self.lease, BerthLease
+            ):
+                raise ValidationError(
+                    _("A BerthProduct must be associated with a BerthLease")
+                )
+            elif isinstance(self.product, WinterStorageProduct) and not isinstance(
+                self.lease, WinterStorageLease
+            ):
+                raise ValidationError(
+                    _(
+                        "A WinterStorageProduct must be associated with a WinterStorageLease"
+                    )
+                )
+        if all([not self.product, not self._product_object_id, not self.base_price]):
+            raise ValidationError(
+                _("Order must have either product object or price value")
+            )
+        if self.lease and self.lease.customer != self.customer:
+            raise ValidationError(
+                _("The lease provided belongs to a different customer")
+            )
+
+        super().clean()
```

Comment:
no need to call `super()` in clean, since it is an empty hook on the parent model ([ref](https://github.com/django/django/blob/master/django/db/models/base.py#L990))

Reply:
I thought about that as well 🤔 it's not really necessary, but I left it in case we would, at some point in the future, do some validation on any parent model. But since it probably won't be the case, I'll get rid of it.

Reply:
👀